apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,483 @@
+
+        x_layer = self.experiment.ms[self.measurement_name].X[layer_name]
+        if not isinstance(x_layer, SOMASparseNDArray):
+            raise TypeError("Dense array unsupported")
+
+        self._load_joinids()
+        return x_layer.read((self._joinids["obs"], self._joinids["var"]))
+
+    def obsp(self, layer: str) -> SparseNDArrayRead:
+        """Return an ``obsp`` layer as a SparseNDArrayRead"""
+        return self._axisp_inner("obs", layer)
+
+    def varp(self, layer: str) -> SparseNDArrayRead:
+        """Return an ``varp`` layer as a SparseNDArrayRead"""
+        return self._axisp_inner("var", layer)
+
+    def _read_axis_dataframe(
+        self,
+        axis: Literal["obs", "var"],
+        axis_df: SOMADataFrame,
+        axis_query: AxisQuery,
+        *,
+        column_names: Optional[Sequence[str]],
+    ) -> pa.Table:
+        """
+        Private. Read the specified axis. Will load and save the resulting soma_joinids for the
+        axis, if they are not already known.
+        """
+        need_joinids = self._joinids[axis] is None
+        query_columns = column_names
+        if (
+            need_joinids
+            and column_names is not None
+            and "soma_joinid" not in column_names
+        ):
+            query_columns = ["soma_joinid"] + list(column_names)
+
+        arrow_table = axis_df.read(
+            ids=axis_query.coords,
+            value_filter=axis_query.value_filter,
+            column_names=query_columns,
+        ).concat()
+
+        if need_joinids:
+            self._joinids[axis] = arrow_table.column("soma_joinid").combine_chunks()
+        assert self._joinids[axis] is not None
+
+        if column_names is not None:
+            arrow_table = arrow_table.select(column_names)
+        return arrow_table
+
+    def _read_both_axes(
+        self,
+        column_names: AxisColumnNames,
+    ) -> Tuple[pa.Table, pa.Table]:
+        """Private. Read both axes in its entirety, ensure that soma_joinid is retained."""
+        futures = (
+            self._threadpool.submit(
+                self._read_axis_dataframe,
+                "obs",
+                self.experiment.obs,
+                self._matrix_axis_query["obs"],
+                column_names=column_names.get("obs", None),
+            ),
+            self._threadpool.submit(
+                self._read_axis_dataframe,
+                "var",
+                self.experiment.ms[self.measurement_name].var,
+                self._matrix_axis_query["var"],
+                column_names=column_names.get("var", None),
+            ),
+        )
+        concurrent.futures.wait(futures)
+        obs_table, var_table = (f.result() for f in futures)
+        return obs_table, var_table
+
+    def read(
+        self,
+        X_name: str,
+        *,
+        use_position_indexing: bool = False,
+        column_names: Optional[AxisColumnNames] = None,
+        X_layers: Optional[Sequence[str]] = None,
```

Comment:
Making this default an empty tuple...
```suggestion
        X_layers: Sequence[str] = (),
```